### PR TITLE
Fix webflux integration to not rely in active span

### PR DESCRIPTION
**Pull request by mar-kolya**
Netty event loop instrumentation doesn't allow us to attribute work to
correct span so we have to maintain that manually.

### CODE REVIEW

File: dd-java-agent/instrumentation/spring-webflux/src/test/groovy/SingleThreadedSpringWebfluxTest.groovy [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.boot.test.context.TestConfiguration
+import org.springframework.boot.web.embedded.netty.NettyReactiveWebServerFactory
+import org.springframework.boot.web.embedded.netty.NettyServerCustomizer
+import org.springframework.context.annotation.Bean
+import reactor.ipc.netty.resources.LoopResources
+
+/**
+ * Run all Webflux tests under netty event loop having only 1 thread.
+ * Some of the bugs are better visible in this setup because same thread is reused
+ * for different requests.
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/spring-webflux/src/test/groovy/SingleThreadedSpringWebfluxTest.groovy]: Nice idea!

File: dd-java-agent/instrumentation/spring-webflux/src/test/groovy/dd/trace/instrumentation/springwebflux/SpringWebFluxTestApplication.groovy [hunk 2]
```diff
@@ -1,12 +1,6 @@
-package datadog.trace.instrumentation.springwebflux
+package dd.trace.instrumentation.springwebflux
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/spring-webflux/src/test/groovy/dd/trace/instrumentation/springwebflux/SpringWebFluxTestApplication.groovy]: Does the package make a difference?

File: dd-java-agent/instrumentation/spring-webflux/src/test/groovy/dd/trace/instrumentation/springwebflux/SpringWebFluxTestApplication.groovy [hunk 3]
```diff
@@ -74,10 +48,20 @@ class SpringWebFluxTestApplication {
...
+    }).andRoute(GET("/greet-delayed"), new HandlerFunction<ServerResponse>() {
+      @Override
+      Mono<ServerResponse> handle(ServerRequest request) {
+        return greetingHandler.defaultGreet().delayElement(Duration.ofMillis(100))
+      }
+    }).andRoute(GET("/greet-failfast/{id}"), new HandlerFunction<ServerResponse>() {
+      @Override
+      Mono<ServerResponse> handle(ServerRequest request) {
+        throw new RuntimeException("bad things happen")
+      }
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/spring-webflux/src/test/groovy/dd/trace/instrumentation/springwebflux/SpringWebFluxTestApplication.groovy]: This could possibly be rewritten (assuming groovy allows it) as:

```groovy
.andRoute(GET("/greet-failfast/{id}")) { ServerRequest request ->
    throw new RuntimeException("bad things happen")
}.andRoute(...
```

File: dd-java-agent/instrumentation/spring-webflux/src/test/groovy/SpringWebfluxTest.groovy [hunk 4]
```diff
@@ -39,73 +51,35 @@ class SpringWebfluxTest extends AgentTestRunner {
...
-            "$Tags.PEER_HOSTNAME.key" "localhost"
-            "$Tags.PEER_PORT.key" Integer
-            "$Tags.HTTP_METHOD.key" "GET"
-            "$Tags.HTTP_STATUS.key" 200
-            "$Tags.HTTP_URL.key" url
-            defaultTags()
+          if (annotatedMethod == null) {
+            // Functional API
+            resourceNameContains(SPRING_APP_CLASS_ANON_NESTED_CLASS_PREFIX, ".handle")
+            operationNameContains(SPRING_APP_CLASS_ANON_NESTED_CLASS_PREFIX, ".handle")
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/spring-webflux/src/test/groovy/SpringWebfluxTest.groovy]: It's hard to follow all the changes in this test here.  I guess I'll trust you on it.

File: dd-java-agent/instrumentation/spring-webflux/src/test/java/dd/trace/instrumentation/springwebflux/RedirectComponent.java [hunk 5]
```diff
@@ -9,6 +9,7 @@
 import org.springframework.web.reactive.function.server.RouterFunction;
 import org.springframework.web.reactive.function.server.ServerResponse;
 
+// Need to keep this in Java becauswe groovy creates crazy proxies around lambdas
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/spring-webflux/src/test/java/dd/trace/instrumentation/springwebflux/RedirectComponent.java]: becauswe -> because

File: dd-java-agent/instrumentation/spring-webflux/src/test/groovy/dd/trace/instrumentation/springwebflux/SpringWebFluxTestApplication.groovy (hunk 2 above)
Comment (mar-kolya) [reply dd-java-agent/instrumentation/spring-webflux/src/test/groovy/dd/trace/instrumentation/springwebflux/SpringWebFluxTestApplication.groovy]: Yes, `datadog.*` doesn't get instrumented - and we need that to test `@Trace` is working.

Comment (mar-kolya) [reply dd-java-agent/instrumentation/spring-webflux/src/test/groovy/dd/trace/instrumentation/springwebflux/SpringWebFluxTestApplication.groovy]: It took me fair amount of time of banging my head against walls to figure this one out :)

File: dd-java-agent/instrumentation/spring-webflux/src/test/groovy/dd/trace/instrumentation/springwebflux/SpringWebFluxTestApplication.groovy (hunk 3 above)
Comment (mar-kolya) [reply dd-java-agent/instrumentation/spring-webflux/src/test/groovy/dd/trace/instrumentation/springwebflux/SpringWebFluxTestApplication.groovy]: The problem is that Groovy seems to be wrapping things into proxies and that makes testing for expected class names harder.

File: dd-java-agent/instrumentation/spring-webflux/src/main/java8/datadog/trace/instrumentation/springwebflux/DispatcherHandlerOnSuccessOrError.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+  public DispatcherHandlerOnSuccessOrError(final ServerWebExchange exchange) {
+    this.exchange = exchange;
+  }
+
+  @Override
+  public void accept(final U object, final Throwable throwable) {
+    // Closing span here means it closes after Netty span which may not be ideal.
+    // We could instrument HandlerFunctionAdapter instead, but this would mean we
+    // would not account for time spent sending request.
```
[the remark this answers is not in the document]
Comment (mar-kolya) [reply dd-java-agent/instrumentation/spring-webflux/src/main/java8/datadog/trace/instrumentation/springwebflux/DispatcherHandlerOnSuccessOrError.java]: I do not think that difference can be large - I think it happens right after read is committed, so nothing large here. It's just traces that look weird with Webflux span closing later than netty span.

File: dd-java-agent/instrumentation/spring-webflux/src/test/groovy/SpringWebfluxTest.groovy (hunk 4 above)
Comment (mar-kolya) [reply dd-java-agent/instrumentation/spring-webflux/src/test/groovy/SpringWebfluxTest.groovy]: I think all preexisting tests are still there. I've added some more tests on top of that.
